### PR TITLE
RFC: Change fastmath to use dispatch instead of dictionary lookup

### PR DESCRIPTION
**Pull request by simonbyrne**
This changes `@fastmath` to use dispatch instead of a dictionary lookup of symbols: this avoids problems whereby a user could use the symbol for another purpose locally, e.g.

``` julia
sin = 2.0
@fastmath  1.0 + sin
```

It is also easily extensible:

``` julia
Base.FastMath.fastmath(::typeof(foo)) = fast_foo
```

will define `fast_foo` as a fast version of `foo` (perhaps a macro would be a good idea here)

cc @eschnett

### CODE REVIEW

File: test/fastmath.jl [hunk 1]
```diff
@@ -144,6 +144,7 @@ end
 for T in (Complex64, Complex128, Complex{BigFloat})
     half = (1+1im)/T(2)
     third = (1-1im)/T(3)
+    println("$T")
```
Comment (tkelman) [inline test/fastmath.jl]: was it intermittent or did adding println make it go away?

Comment (simonbyrne) [inline test/fastmath.jl]: It appears that `println` makes it go away, so I'm not sure what the deal is. The log is here:
https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.5816/job/2lbl9vhx3yfo6x5i
(it only appeared on 64-bit Windows).

File: base/fastmath.jl [hunk 2]
```diff
@@ -25,107 +25,92 @@ export @fastmath
...
-    make_fastmath(esc(expr))
+macro fastmath(x)
+    x
 end
 
+macro deffastmath(ex)
+
+    quote
+        if Base.FastMath.fastmath(f) == f
+            f
```
Comment (tkelman) [inline base/fastmath.jl]: been writing in python lately?

Comment (simonbyrne) [inline base/fastmath.jl]: that wasn't meant to make it in...

**Comment by tkelman** [comment]
appveyor failure looks real here

**Comment by simonbyrne** [comment]
Yeah, same issue as before. Probably triggering some weird floating point exception

**Comment by simonbyrne** [comment]
@vtjnash @yuyichao Any idea what is going on here? I'm consistently getting this error on 64-bit build Appveyor build, which goes away when I add a `println` to figure out what is causing it:

```julia
This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.
signal (22): SIGABRT
while loading C:\projects\julia\test\fastmath.jl, in expression starting on line 144
crt_sig_handler at C:/projects/julia/src\signals-win.c:90
raise at C:\windows\system32\msvcrt.dll (unknown line)
abort at C:\windows\system32\msvcrt.dll (unknown line)
_wassert at /usr/src/debug/mingw64-x86_64-runtime-4.0.6-1/misc\wassert.c:54
_assert at /usr/src/debug/mingw64-x86_64-runtime-4.0.6-1/misc\wassert.c:30
ZN4llvm9DIBuilder17createCompileUnitEjNS_9StringRefES1_S1_bS1_jS1_NS0_17DebugEmissionKindEyb at C:\projects\julia\usr\bin\LLVM-3.7.dll (unknown line)
emit_function at C:/projects/julia/src\codegen.cpp:4092
jl_compile_linfo at C:/projects/julia/src\codegen.cpp:810
jl_compile_for_dispatch at C:/projects/julia/src\gf.c:1309
jl_call_method_internal at C:/projects/julia/src\julia_internal.h:180 [inlined]
jl_apply_generic at C:/projects/julia/src\gf.c:1928
do_call at C:/projects/julia/src\interpreter.c:66
eval at C:/projects/julia/src\interpreter.c:205
jl_toplevel_eval_flex at C:/projects/julia/src\toplevel.c:619 [inlined]
jl_toplevel_eval at C:/projects/julia/src\toplevel.c:641
jl_toplevel_eval_in_warn at C:/projects/julia/src\builtins.c:590
eval at .\boot.jl:234
unknown function (ip: 00000000041FFDF5)
jl_call_method_internal at C:/projects/julia/src\julia_internal.h:185 [inlined]
jl_apply_generic at C:/projects/julia/src\gf.c:1928
macro expansion; at C:\projects\julia\test\fastmath.jl:156 [inlined]
anonymous at .\<missing> (unknown line)
unknown function (ip: 000000000D299933)
jl_call_method_internal at C:/projects/julia/src\julia_internal.h:185 [inlined]
jl_toplevel_eval_flex at C:/projects/julia/src\toplevel.c:630
jl_parse_eval_all at C:/projects/julia/src\ast.c:735
include_string at .\loading.jl:471
unknown function (ip: 0000000004384825)
jl_call_method_internal at C:/projects/julia/src\julia_internal.h:185 [inlined]
jl_apply_generic at C:/projects/julia/src\gf.c:1928
include_from_node1 at .\loading.jl:521
unknown function (ip: 0000000004216401)
jl_call_method_internal at C:/projects/julia/src\julia_internal.h:185 [inlined]
jl_apply_generic at C:/projects/julia/src\gf.c:1928
macro expansion at .\util.jl:232 [inlined]
runtests at C:\projects\julia\test\testdefs.jl:7
unknown function (ip: 000000000D278C66)
jl_call_method_internal at C:/projects/julia/src\julia_internal.h:185 [inlined]
jl_apply_generic at C:/projects/julia/src\gf.c:1928
jl_apply at C:/projects/julia/src\julia.h:1364 [inlined]
jl_f__apply at C:/projects/julia/src\builtins.c:547
#619 at .\multi.jl:1421
run_work_thunk at .\multi.jl:1001
macro expansion at .\multi.jl:1421 [inlined]
#618 at .\event.jl:68
jl_call_method_internal at C:/projects/julia/src\julia_internal.h:185 [inlined]
jl_apply_generic at C:/projects/julia/src\gf.c:1928
jl_apply at C:/projects/julia/src\julia.h:1364 [inlined]
start_task at C:/projects/julia/src\task.c:259
Allocations: 3663421 (Pool: 3662405; Big: 1016); GC: 4
Assertion failed!
Program: C:\projects\julia\usr\bin\julia.exe
File: /home/Tony/julia/deps/srccache/llvm-3.7.1/lib/IR/DIBuilder.cpp, Line 145
```

**Comment by oscardssmith** [comment]
This looks like a cool idea. Is it worth re-doing?

**Comment by simonbyrne** [comment]
Possibly, though it would probably make more sense to use Cassette for this now.

**Pull request closed without merging**